pricegraph/src/graph/shortest_paths.rs

```diff
@@ -64,14 +93,38 @@ where
 
     /// Returns shortest path from source to destination node, if a path exists.
     pub fn path_to(&self, dest: G::NodeId) -> Option<Path<G::NodeId>> {
+        let mut path;
         let mut current = dest;
-        let mut path = Vec::with_capacity(self.predecessor_store.predecessors.len());
-        while current != self.source {
-            path.push(current);
-            current = self.predecessor_store.predecessors[self.graph.to_index(current)]?;
-        }
-        path.push(self.source);
+        path = match &self.predecessor_store {
+            PredecessorStore::Unbounded(predecessors, _) => {
+                path = Vec::with_capacity(predecessors.len());
+                while current != self.source {
+                    path.push(current);
+                    current = predecessors[self.graph.to_index(current)]?;
+                }
+                path
+            }
+            PredecessorStore::Bounded(predecessors_at_step, _) => {
+                let steps = predecessors_at_step.len();
+                path = Vec::with_capacity(steps);
+                let mut step = steps;
+                loop {
+                    if step == 0 {
+                        return None;
+                    }
+                    step -= 1;
+                    if let Some(pred) = predecessors_at_step[step][self.graph.to_index(current)] {
+                        path.push(current);
+                        if pred == self.source {
+                            break path;
+                        }
+                        current = pred;
+                    }
+                }
+            }
```

Comment:
In the description you say:
> The predecessors at step s represent the edge to take to reach the origin in a shortest path of length exactly s.

But then shouldn't the algorithm to find the shortest path iterate over all predecessor lists, starting at the lowest `s` and for each one do the same operation as we do in the `PredecessorStore::Unbounded` case? We have to check whether there is a path at length 0, if not then length 1, if not then length 2, ...
Whereas this algorithm starts at the highest `s`, takes one predecessor and decrements s. I don't understand this. How can it be correct to mix predecessors from different `s`.

Reply:
Looking at this more I don't understand why need to keep the vector of predecessors at each step. Couldn't we keep the one vector as usual but just stop bellman ford after s steps?

Reply:
I think the issue is that in the nth step we might concatenate a paths that is already n edges long (because it was earlier in the loop) with another node (thus making it n+1 edge long). The correct implementation uses the previous vector and thus connects D with distance 4.

Therefore we look at the last predecessor vector, for which are are guaranteed to only have n-1 hops long paths.

E.g.
```
A -- 1 --> B --1--> C --1-->D
|                   ^
\-------3----------/
```

and hops two. In the first step from A we reach node B (1) and node C (3). In the second step we learn how to reach node C in 2 and thus node D in three 3. However the path from A to D in three requires 3 hops (more than we specified).

This might be a nice test case.

Reply:
This is exactly the reason, we might return paths that are in fact longer than the specified number of hops if we do not create a predecessor vector per hop.

Reply:
I have another question for the algorithm here. Would it be possible to instead of storing on predecessor vector for each hop, store a single additional vector that contains for each node the number of hops to the start node. So we would have `predecessors: Vec<PredecessorVec<G>>, weight_distances: Vec<EdgeWeight>, hops_distances: Vec<usize>`. 
Then in bellman_ford we only update the predecessor of a node if we see that the hops doesn't go over the max number of hops.
I haven't fully thought this true, just an idea I had.

Reply:
An issue I see is that it could update the predecessor of a predecessor of the longest path, making the longest path potentially go over the hop limit.

Reply:
A reason to use a predecessor vector for each hop is that finding shortest paths for different nodes may require you to know the shortest path from the same node for a different number of (remaining) hops.
Example (like the one by Felix, but with an extra node at the end):
```
 0 -10.0-> 1 --1.0-> 2 --1.0-> 3
 |         ∧
1.0        |
 v         |
 4 -------1.0
```
Assume `hops = 3`.
Then `shortestWeight(0->3) = weight(0->1->2->3) = 12`. This is using the fact that the shortest path from 0 to 1 with at most _one_ hop has weight 10.
On the other hand, `shortestWeight(0->2) = weight(0->1->4->2) = 3`. Here the shortest path from 0 to 1 with at most _two_ hops has weight 1.
This example could be made arbitrarily more complicated so that you would need the weight of the shortest path from 0 to 1 for any number of steps.